File: cpp/set_matrix_zeroes.cc
/*
 * Filename: set_matrix_zeroes.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Sat 01 Aug 2015 07:12:37 PM EDT
 */

#include <iostream>
using namespace std;

class Solution {
	public:
		void setZeroes(vector<vector<int>>& matrix) {
			if (!matrix.size() || !matrix[0].size()) {
				return;
			}

			int m = matrix.size();
			int n = matrix[0].size();
			bool row = false;
			bool col = false;

			if (!matrix[0][0]) {
				row = true;
				col = true;
			} else {
				for (vector<int>::size_type i = 1; i < m; i++) {
					if (!matrix[i][0]) {
						col = true;
						break;
					}
				}

				for (vector<int>::size_type j = 1; j < n; j++) {
					if (!matrix[0][j]) {
						row = true;
						break;
					}
				}
			}

			for (vector<int>::size_type i = 1; i < m; i++) {
				for (vector<int>::size_type j = 1; j < n; j++) {
					if (matrix[i][j]) {
						continue;
					}

					matrix[i][0] = 0;
					matrix[0][j] = 0;
				}
			}

			for (vector<int>::size_type i = 1; i < m; i++) {
				if (!matrix[i][0]) {
					for (vector<int>::size_type j = 1; j < n; j++) {
						matrix[i][j] = 0;
					}
				}
			}

			for (vector<int>::size_type j = 1; j < n; j++) {
				if (!matrix[0][j]) {
					for (vector<int>::size_type i = 1; i < m; i++) {
						matrix[i][j] = 0;
					}
				}
			}

			if (col) {
				for (vector<int>::size_type i = 0; i < m; i++) {
					matrix[i][0] = 0;
				}
			}

			if (row) {
				for (vector<int>::size_type j = 0; j < n; j++) {
					matrix[0][j] = 0;
				}
			}
		}
};

File: java/SetMatrixZeroes.java
/*
 * Filename: SetMatrixZeroes.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Wed 20 May 2015 01:57:34 PM EDT
 */

public class Solution {
	public void setZeroes(int[][] matrix) {
		if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
			return;
		}

		int m = matrix.length;
		int n = matrix[0].length;
		boolean row0 = false;
		boolean col0 = false;

		for (int i = 0; i < m; i++) {
			if (matrix[i][0] == 0) {
				col0 = true;
				break;
			}
		}

		for (int j = 0; j < n; j++) {
			if (matrix[0][j] == 0) {
				row0 = true;
				break;
			}
		}

		for (int i = 1; i < m; i++) {
			for (int j = 1; j < n; j++) {
				if (matrix[i][j] == 0) {
					matrix[0][j] = 0;
					matrix[i][0] = 0;
				}
			}
		}

		for (int i = 1; i < m; i++) {
			if (matrix[i][0] == 0) {
				for (int j = 1; j < n; j++) {
					matrix[i][j] = 0;
				}
			}
		}

		for (int j = 1; j < n; j++) {
			if (matrix[0][j] == 0) {
				for (int i = 1; i < m; i++) {
					matrix[i][j] = 0;
				}
			}
		}

		if (col0) {
			for (int i = 0; i < m; i++) {
				matrix[i][0] = 0;
			}
		}

		if (row0) {
			for (int j = 0; j < m; j++) {
				matrix[0][j] = 0;
			}
		}
	}
}
